src/mutex.cpp

```diff
@@ -4,51 +4,55 @@
 #include "syscallwrappers.h"
 
 
+static const uint32_t LOCK_FREE = 0;
+static const uint32_t LOCK_ACQUIRED = 1;
+static const uint32_t LOCK_ACQUIRED_WAITERS_QUEUED = 2;
+
 /*
- * Mutex
+ * Mutex has following fields:
+ *   futex: The field for all futex operations
+ *   type:  Lock type.
+ *   owner: Used for recursive locks and to check for deadlocks.
+ *   count: Record recursive lock acquisition count.
  */
 
 extern "C" void
 DmtcpMutexInit(DmtcpMutex *mutex, DmtcpMutexType type)
 {
   mutex->type = type;
+  mutex->futex = 0;
   mutex->owner = 0;
   mutex->count = 0;
 }
 
 
+ /*
+ * If futex is 0 (not acquired), set to 1 (acquired with no waiters) and return.
+ * Otherwise, ensure that it is >1 (acquired, possibly with waiters) and then
+ * block until we acquire the lock, at which point futex will still be >1.  The
+ * lock is always acquired on return.
+ */
 extern "C" int
 DmtcpMutexLock(DmtcpMutex *mutex)
 {
-  pid_t owner = 1;
-
-  if (mutex->type != DMTCP_MUTEX_LLL) {
-    owner = dmtcp_gettid();
-
-    if (mutex->owner == owner) {
-      if (mutex->type == DMTCP_MUTEX_RECURSIVE) {
-        JASSERT(mutex->count + 1 != 0);
-        mutex->count++;
-        return 0;
-      }
-      return EDEADLK;
-    }
+  if (DmtcpMutexTryLock(mutex) == 0) {
+    return 0;
   }
 
-  while (1) {
-    uint32_t waitVal =
-      __sync_val_compare_and_swap((uint32_t *)&mutex->owner, 0, owner);
-
-    if (waitVal == 0) {
-      // We successfully acquired the lock.
-      break;
+  do {
+    int oldval = __sync_val_compare_and_swap(&mutex->futex,
+                                              LOCK_ACQUIRED,
+                                              LOCK_ACQUIRED_WAITERS_QUEUED);
+    if (oldval != LOCK_FREE) {
+      int s = futex_wait(&mutex->futex, LOCK_ACQUIRED_WAITERS_QUEUED);
+      JASSERT(s != -1 || errno == EAGAIN)(JASSERT_ERRNO);
```

Comment:
I believe you want to add a test for EINTR: `(s != -1 || errno == EAGAIN || errno == EINTR)`.

Reply:
Fixed.